src/wasm/wasm-validator.cpp

```diff
@@ -2446,7 +2446,79 @@ void FunctionValidator::visitTry(Try* curr) {
 }
 
 void FunctionValidator::visitTryTable(TryTable* curr) {
-  // TODO
+  shouldBeTrue(
+    getModule()->features.hasExceptionHandling(),
+    curr,
+    "try_table requires exception-handling [--enable-exception-handling]");
+  if (curr->type != Type::unreachable) {
+    shouldBeSubType(curr->body->type,
+                    curr->type,
+                    curr->body,
+                    "try_table's type does not match try_table body's type");
+  } else {
+    shouldBeEqual(curr->body->type,
+                  Type(Type::unreachable),
+                  curr,
+                  "unreachable try_table must have unreachable try_table body");
+  }
+
+  shouldBeEqual(curr->catchTags.size(),
+                curr->catchDests.size(),
+                curr,
+                "the number of catch tags and catch destinations do not match");
+  shouldBeEqual(curr->catchTags.size(),
+                curr->catchRefs.size(),
+                curr,
+                "the number of catch tags and catch refs do not match");
+  shouldBeEqual(curr->catchTags.size(),
+                curr->sentTypes.size(),
+                curr,
+                "the number of catch tags and sent types do not match");
+
+  Type exnref = Type(HeapType::exn, Nullable);
+  for (Index i = 0; i < curr->catchTags.size(); i++) {
+    Name tagName = curr->catchTags[i];
+    if (!tagName) {
+      continue;
+    }
+
+    // Check tag validity
+    auto* tag = getModule()->getTagOrNull(tagName);
+    if (!shouldBeTrue(tag != nullptr, curr, "")) {
+      getStream() << "catch's tag name is invalid: " << tagName << "\n";
+    } else if (!shouldBeEqual(tag->sig.results, Type(Type::none), curr, "")) {
+      getStream()
+        << "catch's tag (" << tagName
+        << ") has result values, which is not allowed for exception handling";
+    }
+
+    // Check types in sentTypes is valid
+    auto tagType = tag->sig.params;
+    auto sentType = curr->sentTypes[i];
+    if (tagType.size() == 0 && !curr->catchRefs[i]) {
+      if (shouldBeTrue(sentType.size() == 1, curr, "")) {
+        shouldBeEqual(sentType,
+                      Type(Type::none),
```

Comment:
What is happening here? `sentType` cannot both have size 1 and be equal to `Type::none`, since `Type::none` has size 0.

Reply:
So in case the sent type is `none`, the `sentTypes` will look like `{Type(Type::none)}`, which is of size 1, no?

To make it contain the sent types that can be used directly at `branch-utils.h`, in case there is no sent values, we add a single `Type::none`:
https://github.com/WebAssembly/binaryen/blob/eb5666eeb0a2b74314e1aaf27e57193cb8fd4a99/src/wasm/wasm.cpp#L927
Usage in `branch-utils.h`
https://github.com/WebAssembly/binaryen/blob/eb5666eeb0a2b74314e1aaf27e57193cb8fd4a99/src/ir/branch-utils.h#L78-L84

Reply:
We can make it `{}` and `branch-utils.h`'s `operateOnScopeNameUsesAndSentTypes` to return `Type::none` in case there is an empty type, but that adds one more `if` and `operateOnScopeNameUsesAndSentTypes` runs more times than the routine popuating `sentTypes` when reading a binary.

Reply:
Sorry, nevermind all above. I didn't know (Type::none).size() returns 0. Will fix.